numba/np/arraymath.py

```diff
@@ -3674,107 +3674,101 @@ def bincount_impl(a, weights=None, minlength=0):
     return bincount_impl
 
 
-def _searchsorted(func):
-    def searchsorted_inner(a, v, v_last, lo, hi, n):
-        """Perform inner loop of searchsorted (i.e. a binary search).
-
-        This is loosely based on the NumPy implementation in [1]_.
-
-        Parameters
-        ----------
-        a: 1-D array_like
-            The input array.
-        v: array_like
-            The current value to insert into `a`.
-        v_last: array_like
-            The previous value inserted into `a`.
-        lo: int
-            The initial/previous "low" value of the binary search.
-        hi: int
-            The initial/previous "high" value of the binary search.
-        n: int
-            The length of `a`.
-
-
-        .. [1] https://github.com/numpy/numpy/blob/809e8d26b03f549fd0b812a17b8a166bcd966889/numpy/core/src/npysort/binsearch.cpp#L173
-        """  # noqa: E501
-        if np.isnan(v):
-            # Find the first nan (i.e. the last from the end of a,
-            # since there shouldn't be many of them in practice)
-            for i in range(n, 0, -1):
-                if not np.isnan(a[i - 1]):
-                    return i
-            return 0
-
-        if v_last < v:
-            hi = n
-        else:
-            lo = 0
-            hi = hi + 1 if hi < n else n
-
-        while hi > lo:
-            mid = (lo + hi) >> 1
-            if func(a[mid], (v)):
-                # mid is too low => go up
-                lo = mid + 1
+@register_jitable
+def custom_lt(a, b):
+    if np.isnan(b):
+        return not np.isnan(a)
+    return a < b
+
+
+@register_jitable
+def custom_le(a, b):
+    if np.isnan(b):
+        return True
+    return a <= b
+
+
+def _searchsorted(func_1, func_2):
+    def impl(a, v):
+        min_idx = 0
+        max_idx = len(a)
+
+        if v.size == 0:
+            return
+
+        out = np.empty(v.size, np.intp)
+
+        last_key_val = v.flat[0]
+
+        for i in range(v.size):
+            key_val = v.flat[i]
+
+            if func_1(last_key_val, key_val):
+                max_idx = len(a)
             else:
-                # mid is too high, or is a NaN => go down
-                hi = mid
-        return lo
-    return searchsorted_inner
+                min_idx = 0
+                if max_idx < len(a):
+                    max_idx += 1
+                else:
+                    max_idx = len(a)
+
+            last_key_val = key_val
 
+            while min_idx < max_idx:
+                mid_idx = min_idx + ((max_idx - min_idx) >> 1)
```

Comment:
Could add a comment here saying this is to avoid overflow

Reply:
done